crates/merkle-tree/src/storage.rs

```diff
@@ -1,154 +1,26 @@
-use anyhow::Context;
-use bitvec::prelude::Msb0;
+use pathfinder_common::trie::TrieNode;
+use pathfinder_storage::{ClassTrieReader, ContractTrieReader, StorageTrieReader};
 use stark_hash::Felt;
 
-use crate::Node;
-
 /// Read-only storage used by the [Merkle tree](crate::tree::MerkleTree).
 pub trait Storage {
-    type Error: std::error::Error + Send + Sync + 'static;
-
-    fn get(&self, node: &Felt) -> Result<Option<Node>, Self::Error>;
+    fn get(&mut self, node: &Felt) -> anyhow::Result<Option<TrieNode>>;
```

Comment:
The argument against it is that you aren't guaranteed to cache the statement then. As in, it will recreate the statement every `get`. And since this operation occurs many times it would be quite bad if it isn't cached.